include/openPMD/backend/Attribute.hpp

```diff
@@ -42,23 +46,154 @@ namespace openPMD
                             float, double, long double,
                             std::string,
                             std::vector< char >,
-                            std::vector< int16_t >,
-                            std::vector< int32_t >,
-                            std::vector< int64_t >,
+                            std::vector< short >,
+                            std::vector< int >,
+                            std::vector< long >,
+                            std::vector< long long >,
                             std::vector< unsigned char >,
-                            std::vector< uint16_t >,
-                            std::vector< uint32_t >,
-                            std::vector< uint64_t >,
+                            std::vector< unsigned short >,
+                            std::vector< unsigned int >,
+                            std::vector< unsigned long >,
+                            std::vector< unsigned long long >,
                             std::vector< float >,
                             std::vector< double >,
                             std::vector< long double >,
                             std::vector< std::string >,
                             std::array< double, 7 >,
                             bool >;
-} // openPMD
+
+template< typename T, typename U, bool isConvertible = std::is_convertible<T, U>::value >
+struct DoConvert;
+
+template< typename T, typename U >
+struct DoConvert<T, U, false>
+{
+    template< typename PV >
+    U operator()( PV )
+    {
+        throw std::runtime_error("getCast: no cast possible.");
+        return U{};
+    }
+};
+
+template< typename T, typename U >
+struct DoConvert<T, U, true>
+{
+    template< typename PV >
+    U operator()( PV pv )
+    {
+        return static_cast< U >( *pv );
+    }
+};
+
+template< typename T, typename U >
+struct DoConvert<std::vector< T >, std::vector< U >, false>
+{
+    static constexpr bool convertible = std::is_convertible<T, U>::value;
+
+    template< typename PV, typename UU = U >
+    auto operator()( PV pv )
+    -> typename std::enable_if< convertible, std::vector< UU > >::type
+    {
+        std::vector< U > u;
+        u.reserve( pv->size() );
+        std::copy( pv->begin(), pv->end(), std::back_inserter(u) );
+        return u;
+    }
+
+    template< typename PV, typename UU = U >
+    auto operator()( PV )
+    -> typename std::enable_if< !convertible, std::vector< UU > >::type
+    {
+        throw std::runtime_error("getCast: no vector cast possible.");
+        return std::vector< U >{};
+    }
+};
+
+/** Retrieve a stored specific Attribute and cast if possible.
+ *
+ * @throw   std::runtime_error if stored object is not static castable to U.
+ * @tparam  U   Type of the object to be casted to.
+ * @return  Copy of the retrieved object, casted to type U.
+ */
+template< typename U >
+inline U
+getCast( Attribute const a )
```

Comment:
Knit-picking here. `DoConvert` returns copies of the cast values anyway, so is there any benefit in the additional copy of the `Attribute`?

Reply:
The only reason I don't put a reference `const &` here is that I can pass temporaries, such as `.dtype()`.

Reply:
Maybe I'm missing something here.
With `Attribute const&`, 
```cpp
auto value = getCast< uint64_t >(magicalTemporary());
```
would not lead to a dangling reference as the value is consumed (i.e. cast and copied) before the destruction of the temporary (as long as that itself is not a reference).

Anyway, let's not waste too much time on this little detail.